### DISCUSSION TITLE
Roadmap and milestones

### DISCUSSION THREAD

**Discussion by aazuspan** [lemma-osu/sknnr#28] [General]
Hey @grovduck, I thought it might be worth using a discussion to talk through how we can start setting up some milestones to help organize our issues/PRs and plan ahead.

Based on [your comment](https://github.com/lemma-osu/scikit-learn-knn-regression/issues/23#issuecomment-1563472861) in #23 and past discussions, I tried to lay out some possible milestones. My thought is that while there's generally an order to the milestones, we can tackle some isolated issues out of order, e.g. example datasets or type checking. 

This is a **very rough** outline, so I propose that if this is helpful we can edit this top-level comment to keep it up to date. 

## 1. Core estimators

- Name: `Core estimators`
- Description: This milestone would conclude with merging branch `fb_add_estimators` into `main`, and would ensure that we have working, well-tested versions of all `KNeighborsRegressor` based estimators--`RawKNNRegressor`, `EuclideanKNNRegressor`, `MahalanobisKNNRegressor`, `GNNRegressor`, and `MSNRegressor`--and all associated transformers.
- Linked milestone: https://github.com/lemma-osu/scikit-learn-knn-regression/milestone/2

## 2. RF-NN

- Name: `RF-NN`
- Description: This would add the last estimator, random-forest based KNN. I imagine this will be a single issue and PR, so I'm not sure whether it makes sense to treat it as a milestone. If there are other features that would make sense to throw in here, we could make this more general.
- Linked milestone: https://github.com/lemma-osu/scikit-learn-knn-regression/milestone/3

## 3. First public release

- Name: `0.1.0` (assuming we want to go with [semver](https://semver.org/))
- Description: This milestone would represent the first public release. At this point, we should have the full set of estimators and transformers with a stable API for all core functionality. We can always add additional features after this point, but breaking changes should be avoided as much as possible. We should also have docstrings, public docs, type checking, licensing, example datasets, and a readme, and publish to PyPI and conda-forge.
- Linked milestone: https://github.com/lemma-osu/scikit-learn-knn-regression/milestone/1

Let me know what you think!

**Comment by grovduck** [comment]
Hey @aazuspan, this looks awesome and thank you for keeping us organized!  If I understand you correctly, rather than this discussion becoming a long thread, we will just want to edit your initial comment, correct?  Do you foresee this serving as a roadmap just until the first public release or afterward as well?  And are you going to create companion milestones or just leave here as a discussion?  I'm game for whatever.

**Reply by aazuspan** [reply]
My thought was we could use this thread to talk through any options, but try to integrate any decisions we make regarding milestones into the top comment, just so we don't have to dig through a long conversation to find the latest developments.

> And are you going to create companion milestones or just leave here as a discussion

Yes, I didn't want to just create a bunch of milestones without talking things through first, so I figured once we settle on a few milestones I can create those and connect the relevant issues.

> Do you foresee this serving as a roadmap just until the first public release or afterward as well?

Not sure! If this ends up being useful, maybe the latter.

**Reply by grovduck** [reply]
> Yes, I didn't want to just create a bunch of milestones without talking things through first, so I figured once we settle on a few milestones I can create those and connect the relevant issues.

I'm definitely +1 on this

**Reply by grovduck** [reply]
I just added the "Core Estimators" [milestone](https://github.com/lemma-osu/scikit-learn-knn-regression/milestone/2) and added a bunch of the work that we've already done and the remaining open issues as well.  Please edit as you see fit.

I think I also figured out how to associate pull requests with issues if they are not on the default branch based on [using the pull request sidebar](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#manually-linking-a-pull-request-to-an-issue-using-the-pull-request-sidebar).  I tried to match pull requests to all closed issues.  We have a few PRs without issues, but those were mostly early commits.

**Reply by aazuspan** [reply]
Awesome, thanks! Seeing the milestone is 70% complete makes it feel like we're actually accomplishing something!

I did come across that sidebar option and have been (inconsistently) using it, but it still doesn't seem to auto-close issues like it would on the default branch. Definitely good to get them linked for future reference though!

**Reply by grovduck** [reply]
> I did come across that sidebar option and have been (inconsistently) using it, but it still doesn't seem to auto-close issues like it would on the default branch. Definitely good to get them linked for future reference though!

Ah, that's good to know.  I had seen a couple of the issues with links and couldn't figure out how you had done that with "closes", "fixes", etc.  We're only 30% away from getting back to that default branch!  That was my bad decision at the start.

**Reply by grovduck** [reply]
@aazuspan, we are getting really close to finishing up the core estimators milestone.  Thanks for all your awesome work!  Should we go ahead and merge back into main after you close #55?

Quick question for this discussion in terms of cleanup.  We are listing individual issues above associated with the milestones, but perhaps we should just link to the milestone there instead (e.g. https://github.com/lemma-osu/scikit-learn-knn-regression/milestone/2)

**Reply by aazuspan** [reply]
Yeah, I think merging back into main makes sense. Exciting!

> We are listing individual issues above associated with the milestones, but perhaps we should just link to the milestone there instead.

Totally! We can keep this discussion around if it's helpful for planning future milestones, but I definitely don't think we need to keep it as a complete record of issues since there's an easier way to handle that, like you mentioned.

**Reply by grovduck** [reply]
Cool, I just linked to the milestones for "core" and "0.1.0".  I've left the one issue intact for RF-NN.

**Reply by aazuspan** [reply]
Does it make sense to call the 2nd milestone something more general like `Complete estimators` and throw #49 in there along with #24? It feels a little weird having a milestone for one PR, even though I'm it'll be a big one... On the other hand, maybe that's a misleading name if there's the potential for us to add more estimators in the future.

Maybe we could also tackle scoring in that milestone, too?

**Reply by grovduck** [reply]
> Does it make sense to call the 2nd milestone something more general like `Complete estimators` and throw #49 in there along with #24? It feels a little weird having a milestone for one PR, even though I'm it'll be a big one...

I think that's a good idea. Definitely RF-NN and the cleanup of `CCA` and `CCorA` are good candidates for a milestone.  There are also a number of estimators in `yaImpute` (called "method" in function `yai`) that I've never dealt with but we could endeavor to add.  I think "Complete Estimators" is clear enough to us what we're trying to do that it works well for a name even if there are other estimators down the road?  I feel like #42 might also be a good candidate for this milestone and then we can make a clean break from `yaImpute`.

> Maybe we could also tackle scoring in that milestone, too?

Can you say a bit more about this?  Given that we now have scoring working both: 1) independently on `X` data that the model was trained on; and 2) new `X` data through the default `KNeighborsRegressor.score`, I've lost track of what we were going to do here.  Was it just that we might introduce different ways to score (i.e. generalized RMSE)?

**Reply by aazuspan** [reply]
Adding regression testing in there sounds like a good idea! I like that 0.1.0 would effectively just be polish at that point.

> I've lost track of what we were going to do here.

Me too! I know we floated a handful of different ideas in our email conversation with Dave, but I'm not sure we reached any conclusions. To avoid derailing this, I opened a new discussion #58.

**Reply by grovduck** [reply]
I've added the "Complete estimators" milestone and added the three issues (#24, #42, #49) in there for now.
